File: yacc/min_caml_asm/ast.cpp
#include "ast.h"

svec_t regs = {"%eax", "%ebx", "%ecx", "%edx", "%esi", "%edi"};
std::set<std::string> allregs = {"%eax", "%ebx", "%ecx", "%edx", "%esi", "%edi"};

bool is_reg(std::string x) {
	return x.c_str()[0] == '%' || x == reg_hp;
}

svec_t
remove_and_uniq(std::set<std::string> xs, svec_t ls) {
	svec_t rc;
	for (auto l : ls) {
		auto it = xs.find(l);
		if (xs.end() != it) continue;
		xs.insert(l);
		rc.push_back(l);		
	}
	return rc;
}

static
svec_t fv_id_or_imm(Id_or_imm* imm) {
	if (auto im = dynamic_cast<V*> (imm))
		return svec_t{im->v};
	return svec_t();
}

static
svec_t fv_exp(Exp* e) {
	if (auto e_ = dynamic_cast<Mov*> (e)) return svec_t{e_->id};
	if (auto e_ = dynamic_cast<Neg*> (e)) return svec_t{e_->id};
	if (auto e_ = dynamic_cast<Save*> (e)) return svec_t{e_->id};
	if (auto e_ = dynamic_cast<Add*> (e))
		return insert<std::string>(svec_t{e_->id},fv_id_or_imm(e_->imm.get()));
	if (auto e_ = dynamic_cast<Sub*> (e))
		return insert<std::string>(svec_t{e_->id},fv_id_or_imm(e_->imm.get()));
	if (auto e_ = dynamic_cast<Ld*> (e))
		return insert<std::string>(svec_t{e_->id},fv_id_or_imm(e_->imm.get()));
	if (auto e_ = dynamic_cast<St*> (e))
		return insert<std::string>(svec_t{e_->id,e_->id2},fv_id_or_imm(e_->imm.get()));
	if (auto e_ = dynamic_cast<Call*> (e)) return e_->ids;
	if (auto e_ = dynamic_cast<IfEq*> (e)) {
		auto v = insert<std::string>(svec_t{e_->id}, fv_id_or_imm(e_->imm.get()));
		auto v2 = insert<std::string>(fv(e_->e1.get()), fv(e_->e2.get()));
		return insert<std::string>(v,remove_and_uniq(std::set<std::string>(), v2));
	}
	if (auto e_ = dynamic_cast<IfLE*> (e)) {
		auto v = insert<std::string>(svec_t{e_->id}, fv_id_or_imm(e_->imm.get()));
		auto v3 = insert<std::string>(fv(e_->e1.get()), fv(e_->e2.get()));
		return insert<std::string>(v, remove_and_uniq(std::set<std::string>(), v3));
	}
	if (auto e_ = dynamic_cast<IfGE*> (e)) {
		auto v = insert<std::string>(svec_t{e_->id}, fv_id_or_imm(e_->imm.get()));
		auto v2 = insert<std::string>(fv(e_->e1.get()), fv(e_->e2.get()));
		return insert<std::string>(v,remove_and_uniq(std::set<std::string>(), v2));
	}
	return svec_t();
}

static
svec_t fv_e(E* e) {
	if (auto e_ = dynamic_cast<Ans*> (e))
		return fv_exp(e_->exp.get());
	if (auto e_ = dynamic_cast<Let*> (e))
		return insert<std::string>(
			fv_exp(e_->exp.get()),
			remove_and_uniq(std::set<std::string>{e_->id}, fv_e(e_->e.get())));
	return svec_t();
}

svec_t fv(E* e) {
	return remove_and_uniq(std::set<std::string>(), fv_e(e));
}

UE concat(UE e1, std::string x, UT t, UE e2) {
	if (auto e = dynamic_cast<Ans*> (std::move(e1).get()))
		return ULet(x, std::move(t), std::move(e->exp), std::move(e2));
	if (auto e = dynamic_cast<Let*> (std::move(e1).get()))
		return ULet(e->id, std::move(e->t), std::move(e->exp), concat(std::move(e->e), x, std::move(t), std::move(e2)));
	return NULL;
}

int align(int i) {
	if (i % 8 == 0) return i;
	else return i + 4;
}

int counter = 0;

std::string genid(std::string s) {
	counter++;
	return s + "." + std::to_string(counter);
}

std::string id_of_typ(T* t) {
	if(dynamic_cast<Unit*>(t)) return "u";
	if(dynamic_cast<Bool*>(t)) return "b";
	if(dynamic_cast<Int*>(t)) return "i";
	if(dynamic_cast<Fun*>(t)) return "f";
	return "";
}

std::string gentmp(T* typ) {
    counter++;
    return std::string("T") + id_of_typ(typ) + std::to_string(counter);
}

File: yacc/min_caml_asm_rust/src/ast.rs
pub const REG_SP : &'static str = "%ebp";
pub const REG_HP : &'static str = "min_caml_hp";
pub const REGS: &'static [&'static str] = &["%eax", "%ebx", "%ecx", "%edx", "%esi", "%edi"];
pub const REG_CL : &'static str = "%edi";
pub fn allregs_find(s:&String) -> bool {
  REGS.contains(&s.as_str())
}

pub fn id_of_typ(t:&T) -> String {
  match *t {
    T::Unit => format!("u"),
    T::Bool => format!("b"),
    T::Int => format!("i"),
    T::Fun(_,_) => format!("f"),
  }
}

#[derive(Clone,Debug,PartialEq)]
pub enum T {
  Unit,
  Bool,
  Int,
  Fun(Vec<T>,Box<T>),
}
#[derive(Clone,Debug,PartialEq)]
pub enum IdOrImm {
  V(String),
  C(i32),
}
#[derive(Clone,Debug,PartialEq)]
pub enum Exp {
  Nop,
  Set(i32),
  SetL(String),
  Mov(String),
  Neg(String),
  Add(String,IdOrImm),
  Sub(String,IdOrImm),
  Ld(String,IdOrImm,i32),
  St(String,String,IdOrImm,i32),
  IfEq(String,IdOrImm,Box<E>,Box<E>),
  IfLE(String,IdOrImm,Box<E>,Box<E>),
  IfGE(String,IdOrImm,Box<E>,Box<E>),
  Call(String,Vec<String>),
  Save(String,String),
  Restore(String),
}
#[derive(Clone,Debug,PartialEq)]
pub enum E {
  Ans(Exp),
  Let(String,T,Exp,Box<E>),
}
#[derive(Clone,Debug,PartialEq)]
pub enum Fundef {
  Fundef(String,Vec<String>,E,T)
}  
#[derive(Clone,Debug,PartialEq)]
pub enum Prog {
  Prog(Vec<Fundef>,E)
}

pub fn fv(e:&E) -> Vec<String> {
  use std::collections::HashSet;
  fn remove_and_uniq(mut xs: HashSet<String>, ls:Vec<String>) -> Vec<String> {
    ls.iter().filter(|&l|
      if !xs.contains(l) {
        xs.insert(l.clone());
        true
      } else {
        false
      }
    ).cloned().collect()
  }
  fn fv_id_or_imm(imm:&IdOrImm) -> Vec<String> {
    match *imm {
      IdOrImm::V(ref v) => vec![v.clone()],
      IdOrImm::C(_) => vec![],
    }
  }
  fn fv_exp(e:&Exp) -> Vec<String> {
    match *e {
      Exp::Nop | Exp::Set(_) | Exp::SetL(_) | Exp::Restore(_) => vec!(),
      Exp::Mov(ref id) | Exp::Neg(ref id) | Exp::Save(ref id, _) => vec!{id.clone()},
      Exp::Add(ref id, ref imm) |  Exp::Sub(ref id, ref imm) | 
      Exp::Ld(ref id, ref imm, _) => {
        let mut v = vec!{id.clone()};
        v.extend(fv_id_or_imm(imm));
        v
      }
      Exp::St(ref id, ref id2, ref imm, _) => {
        let mut v = vec!{id.clone(),id2.clone()};
        v.extend(fv_id_or_imm(imm));
        v
      }
      Exp::Call(_, ref ids) => ids.clone(),
      Exp::IfEq(ref id, ref imm, ref e1, ref e2) |
      Exp::IfLE(ref id, ref imm, ref e1, ref e2) |
      Exp::IfGE(ref id, ref imm, ref e1, ref e2) => {
        let mut v = vec!{id.clone()};
        v.extend(fv_id_or_imm(imm));
        let mut v2 = fv(e1);
        v2.extend(fv(e2));
        v.extend(remove_and_uniq(HashSet::new(), v2));
        v
      }
    }
  }
  fn fv_e(e:&E) -> Vec<String> {
    match *e {
      E::Ans(ref exp) => fv_exp(exp),
      E::Let(ref id, _, ref exp, ref e1) => {
        let mut v = fv_exp(exp);
        let mut set = HashSet::new();
        set.insert(id.clone());
        v.extend(remove_and_uniq(set, fv_e(&e1)));
        v
      }
    }
  }

  remove_and_uniq(HashSet::new(), fv_e(e))
}

use std::cell::Cell;
thread_local!(static F: Cell<i32> = Cell::new(0));
fn gencnt() -> i32 {
  F.with(|f| {
    let cnt = f.get();
    f.set(cnt+1);
    cnt
  })
}

pub fn genid(s:&str) -> String {
  format!("{}{}",s, gencnt())
}

pub fn gentmp(t:&T) -> String {
    genid(&*id_of_typ(t))
}

pub fn seq(e1:Exp, e2:E) -> E {
  E::Let(gentmp(&T::Unit), T::Unit, e1, Box::new(e2))
}
